File: 50-powx-n/powx-n.cpp
class Solution {
public:
    double power(double x, long long n) {
        if (n == 0) {
            return 1;
        }
        if (n < 0) {
            return 1 / power(x, -n);
        }
        
        double temp = power(x, n / 2);
        
        if (n % 2 == 0) {
            return temp * temp;
        } else {
            return temp * temp * x;
        }
    }

    double myPow(double x, int n) {
        return power(x, (long long)n);
    }
};

File: 50-powx-n/powx-n.java
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }
        double ans = 1.0;
        double currentProduct = x;
        while (N > 0) {
            if ((N % 2) == 1) {
                ans = ans * currentProduct;
            }
            currentProduct = currentProduct * currentProduct;
            N /= 2;
        }
        return ans;
    }
}